examples/larger-examples/datalog/subtyping.flix

```diff
@@ -0,0 +1,143 @@
+mod Subtyping {
+
+    type alias Types[r : SchemaRow] = #{ Class(String), Interface(String), ClassExtends(String, String), ClassImplements(String, String), InterfaceExtends(String, String), Method(String, String) | r }
+
+    def open(types: #{ Class(String), Interface(String), ClassExtends(String, String), ClassImplements(String, String), InterfaceExtends(String, String), Method(String, String) }): Types[r] = types
+
+    def subtypeRules(): #{ IsSubtype(String, String), Class(String), Interface(String), ClassExtends(String, String), ClassImplements(String, String), InterfaceExtends(String, String) | r } = #{
```

Comment:
I think it would be nice if we could have a function `subType` or whatever that takes some input as lists and returns a Datalog program which contains a relation SubType where other relations, e.g. Class, ClassExtends, etc. are hidden using local predicates / rho abstraction. Then the types become much smaller. Does it make sense?

Reply:
So instead of all this type aliasing and use of open, I want to use https://doc.flix.dev/local-predicates.html where one function does the heavy computation of SubType and then the rest of the functions simply use it.

Reply:
```
def subType(classes: List[...], classExtends: List[...]): #{SubType(...)} = #(SubType) -> #{ ... }
```

Reply:
Something like this?
```
    pub def subtypes(classes: List[String], interfaces: List[String], classExtensions: List[(String, String)], implementations: List[(String, String)], interfaceExtensions: List[(String, String)], methods: List[String]): #{ IsSubtype(String, String)| r } =
        let cs    = inject classes             into Class;
        let is    = inject interfaces          into Interface;
        let ce    = inject classExtensions     into ClassExtends;
        let impls = inject implementations     into ClassImplements;
        let ies   = inject interfaceExtensions into InterfaceExtends;
        let ms    = inject methods             into Method;
        let rules = #(IsSubtype) -> #{
            IsSubtype(sub, super) :- Class(sub), Class(super), if sub == super.
            IsSubtype(sub, super) :- Interface(sub), Interface(super), if sub == super.
            IsSubtype(sub, super) :- Class(sub), Class(super), ClassExtends(sub, super).
            IsSubtype(sub, super) :- Class(sub), Interface(super), ClassImplements(sub, super).
            IsSubtype(sub, super) :- Interface(sub), Interface(super), InterfaceExtends(sub, super).
            IsSubtype(sub, super) :- IsSubtype(sub, a), IsSubtype(a, super).
        };
        solve cs, is, ce, impls, ies, ms, rules project IsSubtype

```

Reply:
Yes, but the rules look strange. You can just do 

ISSubtype(x, x) :- Class(x)

Reply:
In addition to subtype, I would also like something like HasMethod(className, methodName).

Reply:
But yes, this looks like the right direction. So we have one method that computes some general information and then we have smaller methods that call this big method and answer specify queries.